src/QsCompiler/Compiler/RewriteSteps/AbstractRewriteStepsLoader.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.CompilationBuilder;
+using Microsoft.Quantum.QsCompiler.Diagnostics;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.Quantum.QsCompiler
+{
+    internal abstract class AbstractRewriteStepsLoader : IRewriteStepsLoader
+    {
+        protected readonly Action<Diagnostic>? onDiagnostic;
+        protected readonly Action<Exception>? onException;
+
+        protected AbstractRewriteStepsLoader(Action<Diagnostic>? onDiagnostic = null, Action<Exception>? onException = null)
+        {
+            this.onDiagnostic = onDiagnostic;
+            this.onException = onException;
+        }
+
+        public abstract ImmutableArray<LoadedStep> GetLoadedSteps(CompilationLoader.Configuration config);
+
+        protected Diagnostic LoadError(Uri target, ErrorCode code, params string[] args) => Errors.LoadError(code, args, ProjectManager.MessageSource(target));
+
+        protected Diagnostic LoadWarning(Uri target, WarningCode code, params string[] args) => Warnings.LoadWarning(code, args, ProjectManager.MessageSource(target));
+
+        protected LoadedStep CreateStep(Type type, Uri target, string? outputFolder)
+        {
+            try
+            {
+                var instance = Activator.CreateInstance(type);
+                if (instance is IRewriteStep step)
+                {
+                    return new LoadedStep(step, typeof(IRewriteStep), target, outputFolder);
+                }
+
+                // we also try to load rewrite steps that have been compiled against a different compiler version
+                try
+                {
+                    var interfaceType = type.GetInterfaces().First(t => t.FullName == typeof(IRewriteStep).FullName);
+                    var loadedStep = new LoadedStep(instance, interfaceType, target, outputFolder);
+                    this.onDiagnostic?.Invoke(this.LoadWarning(target, WarningCode.RewriteStepLoadedViaReflection, loadedStep.Name, target.LocalPath));
+                    return loadedStep;
+                }
+                catch
+                {
+                    // we don't log the exception, since it is perfectly possible that we should have ignored this type in the first place
+                    this.onDiagnostic?.Invoke(this.LoadWarning(target, WarningCode.FailedToLoadRewriteStepViaReflection, target.LocalPath));
+                }
+            }
+            catch (Exception ex)
+            {
+                this.onDiagnostic?.Invoke(this.LoadError(target, ErrorCode.CouldNotInstantiateRewriteStep, type.ToString(), target.LocalPath));
+                this.onException?.Invoke(ex);
+            }
+
+            return LoadedStep.Empty;
```

Comment:
Since this step would not really be usable, I would tend to actually make it more obvious that it is not a successfully loaded step by returning null.

Reply:
I actually had it like that in the beginning, but then nullability analysis was added to the project in a separate PR. I wasn't sure whether to make it nullable at that point anymore, so I decided to use the [null object pattern](https://en.wikipedia.org/wiki/Null_object_pattern)

Reply:
done